utils/testtools/crypto.go

```diff
@@ -135,72 +79,138 @@ func CreateCSR(key crypto.PrivateKey) (csr []byte, err error) {
-	tmpDir, err := ioutil.TempDir(workDir, "cert*")
+	caBytes, err := x509.CreateCertificate(rand.Reader, template, parent, publicKey, privateKey)
 	if err != nil {
 		return nil, aoserrors.Wrap(err)
 	}
 
-	defer os.RemoveAll(tmpDir)
+	cert, err := x509.ParseCertificate(caBytes)
+	if err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
 
-	caCertFile := path.Join(tmpDir, "ca.crt")
-	caKeyFile := path.Join(tmpDir, "ca.key")
-	csrFile := path.Join(tmpDir, "unit.csr")
-	csrConfFile := path.Join(tmpDir, "csr.conf")
-	unitCertFile := path.Join(tmpDir, "unit.der")
+	return cert, nil
+}
 
-	if err = ioutil.WriteFile(csrFile, csr, filePerm); err != nil {
-		return nil, aoserrors.Wrap(err)
+// GenerateKeyAndCertificate generate key and certificate.
+func GenerateKeyAndCertificate(
+	template, parent *x509.Certificate, parentPrivateKey *rsa.PrivateKey,
+) (cert *x509.Certificate, privateKey *rsa.PrivateKey, err error) {
+	key, err := rsa.GenerateKey(rand.Reader, 2048) // nolint:gomnd
+	if err != nil {
+		return nil, nil, aoserrors.Wrap(err)
 	}
 
-	if err = ioutil.WriteFile(caCertFile, GetCACertificate(), filePerm); err != nil {
-		return nil, aoserrors.Wrap(err)
+	if parentPrivateKey == nil {
+		parentPrivateKey = key
 	}
 
-	if err = ioutil.WriteFile(caKeyFile, GetCAKey(), filePerm); err != nil {
-		return nil, aoserrors.Wrap(err)
+	if cert, err = GenerateCertificate(template, parent, parentPrivateKey, key.Public()); err != nil {
+		return nil, nil, err
 	}
 
-	if err = ioutil.WriteFile(csrConfFile, []byte(csrConf), filePerm); err != nil {
-		return nil, aoserrors.Wrap(err)
+	return cert, key, nil
+}
+
+// Verify verify chain of certificates.
+func Verify(leafCert *x509.Certificate, roots, intermediates []*x509.Certificate) error {
+	rootPool := x509.NewCertPool()
+	intermediatePool := x509.NewCertPool()
+
+	for _, c := range roots {
+		rootPool.AddCert(c)
 	}
 
-	var out []byte
+	for _, c := range intermediates {
+		intermediatePool.AddCert(c)
+	}
 
-	if out, err = exec.Command("openssl", "req", "-inform", "PEM", "-in",
-		csrFile, "-out", csrFile+".pem").CombinedOutput(); err != nil {
-		return nil, aoserrors.Errorf("message: %s, %s", string(out), err)
+	opts := x509.VerifyOptions{
+		Roots:         rootPool,
+		Intermediates: intermediatePool,
 	}
 
-	if out, err = exec.Command("openssl", "x509", "-req", "-in", csrFile+".pem",
-		"-CA", caCertFile, "-CAkey", caKeyFile, "-extfile", csrConfFile, "-extensions", "ext",
-		"-outform", "PEM", "-out", unitCertFile, "-CAcreateserial", "-days", "3650").CombinedOutput(); err != nil {
-		return nil, aoserrors.Errorf("message: %s, %s", string(out), err)
+	if _, err := leafCert.Verify(opts); err != nil {
+		return aoserrors.Wrap(err)
 	}
 
-	certData, err := ioutil.ReadFile(unitCertFile)
+	return nil
+}
+
+// GenDefaultCARootCertPEMAndKey generate default CA root certificate and key.
+func GenDefaultCARootCertPEMAndKey() (cert *x509.Certificate, key *rsa.PrivateKey, err error) {
```

Comment:
Use either `Gen` or `Generate` prefix for all functions. Do you need `PEM` word in the name of function?

Reply:
Done